File: C#/LocalSearchOptimization/Solvers/LocalDescent.cs
using System;
using System.Collections.Generic;
using LocalSearchOptimization.Components;
using LocalSearchOptimization.Parameters;

namespace LocalSearchOptimization.Solvers
{
    public class LocalDescent : IOptimizationAlgorithm
    {
        private LocalDescentParameters parameters;

        private ISolution currentSolution;

        private List<SolutionSummary> searchHistory;

        private bool stopFlag = false;

        public ISolution CurrentSolution { get => currentSolution; }

        public List<SolutionSummary> SearchHistory { get => searchHistory; }

        public LocalDescent(LocalDescentParameters parameters)
        {
            this.parameters = parameters;
        }

        public IEnumerable<ISolution> Minimize(ISolution startSolution)
        {
            stopFlag = false;
            int iteration = 0;
            DateTime startedAt = DateTime.Now;
            currentSolution = startSolution;
            startSolution.IterationNumber = 0;
            startSolution.TimeInSeconds = 0;
            startSolution.IsCurrentBest = true;
            startSolution.IsFinal = false;
            startSolution.InstanceTag = parameters.Name;
            searchHistory = new List<SolutionSummary>
            {
                new SolutionSummary
                {
                    InstanceTag = parameters.Name,
                    OperatorTag = currentSolution.OperatorTag,
                    IterationNumber = currentSolution.IterationNumber,
                    CostValue = currentSolution.CostValue
                }
            };
            Neighborhood neighborhood = new Neighborhood(currentSolution, parameters.Operators, parameters.Seed);
            bool bestFound;
            do
            {
                bestFound = true;
                foreach (ISolution neighbor in neighborhood.Neighbors)
                {
                    iteration++;
                    if (neighbor.CostValue < currentSolution.CostValue)
                    {
                        currentSolution = neighbor;
                        bestFound = false;
                        if (!parameters.IsSteepestDescent) break;
                    }
                }
                if (!bestFound)
                {
                    currentSolution.IterationNumber = iteration;
                    currentSolution.TimeInSeconds = (DateTime.Now - startedAt).TotalSeconds;
                    currentSolution.IsCurrentBest = true;
                    currentSolution.IsFinal = false;
                    currentSolution.InstanceTag = parameters.Name;
                    searchHistory.Add(new SolutionSummary
                    {
                        InstanceTag = parameters.Name,
                        OperatorTag = currentSolution.OperatorTag,
                        IterationNumber = currentSolution.IterationNumber,
                        CostValue = currentSolution.CostValue
                    });
                    if (parameters.DetailedOutput) yield return currentSolution;
                    neighborhood.MoveToSolution(currentSolution);
                }
            } while (!(bestFound || stopFlag));
            currentSolution.IterationNumber = iteration;
            currentSolution.TimeInSeconds = (DateTime.Now - startedAt).TotalSeconds;
            currentSolution.IsFinal = true;
            yield return currentSolution;
            Console.WriteLine("{0} finished with cost {1} at iteration {2}", parameters.Name, currentSolution.CostValue, iteration);
        }

        public void Stop()
        {
            stopFlag = true;
        }
    }
}

File: java/src/localsearchoptimization/solvers/LocalDescent.java
package localsearchoptimization.solvers;

import localsearchoptimization.components.*;
import localsearchoptimization.parameters.LocalDescentParameters;

import java.util.ArrayList;
import java.util.Arrays;

public class LocalDescent implements OptimizationAlgorithm {

    private LocalDescentParameters parameters;

    private SolutionHandler solutionHandler;

    private Solution currentSolution;

    private ArrayList<SolutionSummary> searchHistory;

    private boolean stopFlag = false;

    public LocalDescent(LocalDescentParameters parameters) {
        this(parameters, null);
    }

    public LocalDescent(LocalDescentParameters parameters, SolutionHandler solutionHandler) {
        this.parameters = parameters;
        this.solutionHandler = solutionHandler;
    }

    @Override
    public Solution minimize(Solution startSolution) {
        stopFlag = false;
        int iteration = 0;
        long startedAt = System.currentTimeMillis();
        currentSolution = startSolution;
        startSolution.setIterationNumber(0);
        startSolution.setElapsedTime(0);
        startSolution.isCurrentBest(true);
        startSolution.isFinal(false);
        startSolution.setInstanceTag(parameters.name);
        searchHistory = new ArrayList<SolutionSummary>(
                Arrays.asList(new SolutionSummary(parameters.name,
                        currentSolution.getOperatorTag(),
                        currentSolution.getIterationNumber(),
                        currentSolution.getCost())));
        Neighborhood neighborhood = new Neighborhood(currentSolution, parameters.operators, parameters.seed);
        boolean bestFound;
        do {
            bestFound = true;
            while (neighborhood.hasNext()) {
                iteration++;
                Solution neighbor = neighborhood.getNext();
                if (neighbor.getCost() < currentSolution.getCost()) {
                    currentSolution = neighbor;
                    bestFound = false;
                    if (!parameters.isSteepestDescent) break;
                }
            }
            if (!bestFound) {
                currentSolution.setIterationNumber(iteration);
                currentSolution.setElapsedTime((System.currentTimeMillis() - startedAt) / 1000.0);
                currentSolution.isCurrentBest(true);
                currentSolution.isFinal(false);
                currentSolution.setInstanceTag(parameters.name);
                searchHistory.add(new SolutionSummary(
                        parameters.name,
                        currentSolution.getOperatorTag(),
                        currentSolution.getIterationNumber(),
                        currentSolution.getCost()
                ));
                if (parameters.isDetailedOutput && solutionHandler != null)
                    solutionHandler.process(currentSolution);
                neighborhood.moveToSolution(currentSolution);
            }
        } while (!(bestFound || stopFlag));
        currentSolution.setIterationNumber(iteration);
        currentSolution.setElapsedTime((System.currentTimeMillis() - startedAt) / 1000.0);
        currentSolution.isFinal(true);
        System.out.printf("\t%1$s finished with getCost %2$s at iteration %3$d, time %4$.2f\n", parameters.name, currentSolution.getCost(), iteration, currentSolution.getElapsedTime());
        if (solutionHandler != null)
            solutionHandler.process(currentSolution);
        return currentSolution;
    }

    @Override
    public Solution getCurrentSolution() {
        return currentSolution;
    }

    @Override
    public ArrayList<SolutionSummary> getSolutionsHistory() {
        return searchHistory;
    }

    @Override
    public void stop() {
        stopFlag = true;
    }
}